gateway/ln-gateway/src/lightning/ldk.rs

```diff
@@ -290,7 +356,50 @@ impl ILnRpcClient for GatewayLdkClient {
         max_delay: u64,
         max_fee: Amount,
     ) -> Result<PayInvoiceResponse, LightningRpcError> {
-        let payment_id = match self.node.bolt11_payment().send(
+        // Subscribe to the `PaymentSuccessful`/`PaymentFailed` events for this invoice.
+        let (sender, receiver) = tokio::sync::oneshot::channel();
+        self.payment_subscriptions
+            .lock()
+            .await
+            .insert(*invoice.payment_hash(), sender);
+
+        // Check if the invoice payment has already been started. This could happen if
+        // the gateway was stopped after the payment was initiated but before it was
+        // completed. It's important that this is done after the subscription is added
+        // to the map, otherwise we might miss the event.
+        let payment_details_or = self
+            .node
+            .list_payments_with_filter(|payment| {
+                if let PaymentKind::Bolt11 { hash, .. } = payment.kind {
+                    hash.0 == invoice.payment_hash().to_byte_array()
+                } else {
+                    false
+                }
+            })
+            .into_iter()
+            .next();
```

Comment:
This PR tries to achieve never paying an invoice twice, can we assert here that the iterator actually only had one item?

Reply:
Good point! Done.